saffron/src/read_proof.rs

```diff
@@ -0,0 +1,381 @@
+//! This module defines the read proof prover and verifier. Given a
+//! query vector q, a vector of data d, and a commitment to this data
+//! C, the prover will return an answer a and a proof that the answers
+//! correspond to the data committed in C at the specified indexes in
+//! the query.
+//!
+//! The folding version is TBD
+//! We call data the data vector that is stored and queried
+//! We call answer the vector such that answer[i] = data[i] * query[i]
+
+use crate::{Curve, CurveFqSponge, CurveFrSponge, ScalarField};
+use ark_ff::{Field, Zero};
+use ark_poly::{
+    univariate::DensePolynomial, EvaluationDomain, Evaluations, Polynomial,
+    Radix2EvaluationDomain as R2D,
+};
+use kimchi::{circuits::domains::EvaluationDomains, curve::KimchiCurve, plonk_sponge::FrSponge};
+use mina_poseidon::FqSponge;
+use poly_commitment::{
+    commitment::{combined_inner_product, BatchEvaluationProof, CommitmentCurve, Evaluation},
+    ipa::{OpeningProof, SRS},
+    utils::DensePolynomialOrEvaluations,
+    PolyComm, SRS as _,
+};
+use rand::{CryptoRng, RngCore};
+use tracing::instrument;
+
+// #[serde_as]
+#[derive(Debug, Clone)]
+// TODO? serialize, deserialize
+pub struct ReadProof {
+    // Commitment to the query vector
+    pub query_comm: Curve,
+    // Commitment to the answer
+    pub answer_comm: Curve,
+    // Commitment of quotient polynomial T (aka t_comm)
+    pub quotient_comm: Curve,
+
+    // Evaluation of data polynomial at the required challenge point
+    pub data_eval: ScalarField,
+    // Evaluation of query polynomial at the required challenge point
+    pub query_eval: ScalarField,
+    // Evaluation of answer polynomial at the required challenge point
+    pub answer_eval: ScalarField,
+
+    // Polynomial commitment’s proof for the validity of returned evaluations
+    pub opening_proof: OpeningProof<Curve>,
+}
+
+#[instrument(skip_all, level = "debug")]
+pub fn prove<RNG>(
+    domain: EvaluationDomains<ScalarField>,
+    srs: &SRS<Curve>,
+    group_map: &<Curve as CommitmentCurve>::Map,
+    rng: &mut RNG,
+    // data is the data that is stored and queried
+    data: &[ScalarField],
+    // data[i] is queried if query[i] ≠ 0
+    query: &[ScalarField],
+    // answer[i] = data[i] * query[i]
+    answer: &[ScalarField],
+    // Commitment to data
+    data_comm: &Curve,
+) -> ReadProof
+where
+    RNG: RngCore + CryptoRng,
+{
+    let mut fq_sponge = CurveFqSponge::new(Curve::other_curve_sponge_params());
+
+    let data_d1 = Evaluations::from_vec_and_domain(data.to_vec(), domain.d1);
+    let data_poly: DensePolynomial<ScalarField> = data_d1.clone().interpolate();
+    let data_comm: PolyComm<Curve> = PolyComm {
+        chunks: vec![*data_comm],
+    };
+
+    let query_d1 = Evaluations::from_vec_and_domain(query.to_vec(), domain.d1);
+    let query_poly: DensePolynomial<ScalarField> = query_d1.clone().interpolate();
+    let query_comm: Curve = srs.commit_non_hiding(&query_poly, 1).chunks[0];
```

Comment:
Follow-up: we're gonna need to hide with a blinder to support the zero polynomial.